File: src/dotnet/Taml.NET/Taml.NET/Parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace TAML
{

	/// <summary>
	/// Read, interpret, and convert TAML to elements that can be coded against
	/// </summary>
	public class Parser
	{
		private static readonly Regex _KeyValuePair = new Regex(@"(?<key>\S[^\t]*)\t+(?<value>\S[^\t]*)");
		private static readonly Regex _SingleValue = new Regex(@"^\t*\S[^\t]*\t*$");

		/// <summary>
		/// The maximum TAML specification version supported by this parser
		/// </summary>
		/// <returns></returns>
		public static Version SupportedSpecVersion => new Version(1,1); 

		/// <summary>
		/// Convert the TAML submitted into a TamlDocument that can be navigated
		/// </summary>
		/// <param name="reader">A StreamReader containing referencing the TAML content to be parsed</param>
		/// <returns>A TamlDocument that represents the string TAML document</returns>
		public static TamlDocument Parse(StreamReader reader)
		{

			var parsedDoc = ReadLines(reader);
			var lines = parsedDoc.Lines;

			var document = new TamlDocument();
			document.ProcessorDirectives = parsedDoc.ProcessorDirectives;

			int currentLevel = 0;

			var stack = new Stack<TamlKeyValuePair>();

			for (var i = 0; i < lines.Count; i++)
			{

				var (indent, currentPair) = lines[i];

				if (indent == 0)
				{
					// line is root element
					document.KeyValuePairs.Add(currentPair!);
					currentLevel = 0;
					continue;
				}

				if (indent > currentLevel)
				{
					stack.Push(lines[i - 1].value);
				}
				else if (indent < currentLevel)
				{
					stack.Pop();
				}

				TamlKeyValuePair? parent = stack.Count > 0 ? stack.Peek() : null;

				if (indent > currentLevel)
				{
					if (parent!.Value is null)
					{
						var newArray = new TamlArray();
						newArray.AppendValue(currentPair);
						parent.Value = newArray;
					}
				}
				else if (parent != null && parent.Value is TamlArray parentArray)
				{
					parentArray.AppendValue(currentPair);
				}

				currentLevel = indent;
			}
			return document;
		}

		public static TamlDocument ParseFile(string fileName)
		{
			var sr = new StreamReader(fileName);
			return Parse(sr);
		}

		private static int CountIntendedTabs(string currentLine)
		{
			int tabs = 0;

			while (currentLine[tabs] == '\t')
			{
				tabs++;
			}

			return tabs;
		}

		private static TamlValue GetPairIfNoValue(TamlKeyValuePair currentPair)
		{
			return currentPair.HasValue && !string.IsNullOrEmpty(currentPair.Value) ? currentPair : new TamlKeyValuePair(currentPair!.Key, null);
		}

		private static ParsedDocument ReadLines(StreamReader reader)
		{

			var directives = new List<TamlKeyValuePair>();
			var lines = new Dictionary<int, (int indent, TamlKeyValuePair value)>();
			int currentLineNumber = 0;

			while (!reader.EndOfStream)
			{
				var rawLine = reader.ReadLine();

				// Handle comments - ignore and read next line
				if (rawLine?[0] == '#') continue;

				var indent = CountIntendedTabs(rawLine);
				var line = rawLine.Trim();
				TamlKeyValuePair? value = null;

				if (string.IsNullOrEmpty(line))
				{
					// empty line we ignore
					currentLineNumber--; // we don't count empty lines
					continue;
				}
				else if (line.Contains('\t'))
				{
					// this is a KeyValuePair
					var match = _KeyValuePair.Matches(line)[0].Groups;

					value = new TamlKeyValuePair(match["key"].Value, new TamlValue(match["value"].Value));
				}
				else
				{
					// this is a single value
					value = new TamlKeyValuePair(line, null);
				}

				if (rawLine?[0] == '!') // This is a processor directive
				{
					value.Key = value.Key.TrimStart('!', ' ');
					directives.Add(value);
				}
				else
				{
					lines.Add(currentLineNumber, (indent, value));
					currentLineNumber++;
				}
			}

			return new ParsedDocument(lines, directives);

		}
	}
}

File: src/JVM/taml4j-parser/src/main/java/com/taml4j/Parser.java
package com.taml4j;

import java.io.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Luca Taddeo - Lucalas
 */
public class Parser {
	private static final Pattern keyValuePair = Pattern.compile("(?<key>\\S[^\\t]*)\\t+(?<value>\\S[^\\t]*)", Pattern.MULTILINE);
	private static final Pattern singleValue = Pattern.compile("^\t*\\S[^\t]*\t*$");


	public static TamlDocument Parse(BufferedReader reader) {
		ParsedDocument parsedDoc = ReadLines(reader);
		Map<Integer, KeyValuePair<Integer, TamlKeyValuePair>> lines = parsedDoc.getLines();

		TamlDocument document = new TamlDocument();
		document.setProcessorDirectives(parsedDoc.getProcessorDirectives());

		int currentLevel = 0;

		Stack<TamlKeyValuePair> stack = new Stack<>();

		int i = 0;
		Integer lastLineKey = null;
		for (Map.Entry<Integer, KeyValuePair<Integer, TamlKeyValuePair>> line : lines.entrySet()) {

			int indent = line.getValue().getKey();
			TamlKeyValuePair currentPair = line.getValue().getValue();
			if (indent == 0)
			{
				// line is root element
				document.getKeyValuePairs().add(currentPair);
				currentLevel = 0;
				lastLineKey = line.getKey();
				continue;
			}

			if (indent > currentLevel)
			{
				stack.push(lines.get(lastLineKey).getValue());
			}
			else if (indent < currentLevel)
			{
				stack.pop();
			}

			TamlKeyValuePair parent = !stack.empty() ? stack.peek() : null;

			if (indent > currentLevel)
			{
				if (parent != null && parent.getValue() == null)
				{
					TamlArray newArray = new TamlArray();
					newArray.appendValue(currentPair);
					parent.setValue(newArray);
				}
			}
			else if (parent != null && parent.getValue() instanceof TamlArray)
			{
				((TamlArray)parent.getValue()).appendValue(currentPair);
			}
			lastLineKey = line.getKey();
			currentLevel = indent;
		}
		return document;
	}

	private static ParsedDocument ReadLines(BufferedReader reader)
	{
		TamlMap directives = new TamlMap();
		Map lines = new HashMap<Integer, KeyValuePair<Integer, TamlKeyValuePair>>();
		int currentLineNumber = 0;
		try {
			String rawLine;
				while ((rawLine = reader.readLine()) != null) {

					// Handle comments - ignore and read next line
					if (rawLine.charAt(0) == '#') continue;

					int indent = CountIntendedTabs(rawLine);
					String line = rawLine.trim();
					TamlKeyValuePair value;

					if (line == null || line.isEmpty())
					{
						// empty line we ignore
						currentLineNumber--; // we don't count empty lines
						continue;
					}
					else if (line != null && line.contains("\t"))
					{
						// this is a KeyValuePair
						Matcher match = keyValuePair.matcher(line);
						match.find();
						value = new TamlKeyValuePair(match.group("key"), new TamlValue(match.group("value")));
					}
					else
					{
						// this is a single value
						value = new TamlKeyValuePair(line, null);
					}

					if (rawLine.charAt(0) == '!') // This is a processor directive
					{
						value.setKey(value.getKey().replaceAll("^*[ !]", ""));
						directives.put(value.getKey(), value.getValue());
					}
				else
					{
						lines.put(currentLineNumber, new KeyValuePair<>(indent, value));
						currentLineNumber++;
					}
				}
		} catch (IOException e) {
			e.printStackTrace();
		}

		return new ParsedDocument(lines, directives);

	}

	private static int CountIntendedTabs(String currentLine)
	{
		int tabs = 0;
		while (currentLine.charAt(tabs) == '\t')
		{
			tabs++;
		}
		return tabs;
	}
}
